enhancements/cluster-logging/cluster-logging-es-pipeline-processing.md

```diff
@@ -0,0 +1,205 @@
+
+However, common data model does not solve another challenge associated with the ask to properly index, store and query any additional (extra) custom data the documents can carry. For example logs can contain project or business domain specific JSON data that is not captured in the common data model and is not shared by all the data sources.
+
+Over the time we have gathered number of requests from customers asking for possibility to process additional data. See <https://issues.redhat.com/browse/LOG-785>.
+
+For the purpose of this proposal we have already investigated various approaches (and their pros and cons) of storing additional JSON data in Elasticsearch indices as part of <https://issues.redhat.com/browse/LOG-842>.
+
+
+### Goals
+
+The high-level goal of this proposal is to lay the groundwork to enable approaches B) and C) as discussed in [LOG-842](https://issues.redhat.com/browse/LOG-842).
+
+Quick comparison of approaches B) and C):
+
+- Approach B) Whenever user needs to store data with a new data model we solve it by pointing this data to new index (or to existing index if there is already one with exactly the same data model). This is very comfy approach for the user except it can lead to very high number of indices and querying experience can be complicated because there are very little or no common conventions of document fields meaning and types.
+
+- Approach C) is on the other side of the spectrum. All data is stored into common index and whenever user needs to store data with a new data model then it is necessary to transform the data to fit existing data model in existing index. This approach requires attention to data model and users may need to invest time and resources to transform new data sources. On the other hand there is minimum additional indices created in ES cluster and data can be queried uniformly because document fields have predictable meaning and types.
+
+It should be pointed out that Elasticsearch pipeline processes can support both cases. Depending on how much users are willing to pay attention to their data model they can start with approach B) and ventually transition to approach C) once they feel more confident about managing their own data model across more log resources.
+
+In this proposal we assume that the approach B is the one we would start enabling first. But nothing is stopping us from enabling the approach C in the future. Though it is fair to say that approach C will be more complicated to implement because it will require management of custom processing pipelines.
+
+We will be successful when:
+
+* An OCP user (or admin only?)  is able to configure specific logs to be pointed into different (non-default) index in Elasticsearch based on the data contained in the document itself.
```

Comment:
this brings up a good point, is this something we will let users configure? or will we simply say "with EO managed Elasticsearch on OpenShift you get x, y, z" ?
@jcantrill @christianvogt @alanconway

Reply:
This should go through some "approval process". Meaning if users are able to do it they **must** understand there is associated cost to this and at some point the cost can be so high that ES cluster stability can be in risk. I did not address this "approval process" in this proposal on purpose - because from ES side perspective this is not important, it all gets to what kind of rule is pushed into pipeline processing, if the rule is to create a new index when specific annotation field is present of has specific value - we do not care.

ATM this proposal is "just" about enabling the use of ES pipeline processors and eventually pointing incoming log into different index (and optionally we can also parse some string fields into JSON along the lines too or do other transformations to the logs before they are indexed).

Reply:
can we reword this then as "We are able to configure" ... ?

Reply:
Done.